VSRAD.Syntax/Core/DocumentFactory.cs

```diff
@@ -92,71 +86,89 @@ public IDocument GetOrCreateDocument(ITextBuffer buffer)
             var lexerParser = GetLexerParser(textDocument.TextBuffer.GetAsmType());
             if (!lexerParser.HasValue) return null;
 
-            var document = creator(lexerParser.Value.Lexer, lexerParser.Value.Parser);
-            ObserveDocument(document, textDocument);
+            IDocument document;
+            if (observe && _documents.TryGetValue(textDocument.FilePath, out var documentPair))
+            {
+                document = documentPair.Value;
+                document.ReplaceDocument(textDocument);
+                OnDocumentDestroy(document);
+            }
+            else
+            {
+                document = creator(lexerParser.Value.Lexer, lexerParser.Value.Parser);
+            }
+
+            if (observe)
+            {
+                textDocument.FileActionOccurred += TextDocumentActionOccurred;
+                _documents.Add(document.Path, new KeyValuePair<ITextDocument, IDocument>(textDocument, document));
+            }
 
             return document;
         }
 
-        private void ObserveDocument(IDocument document, ITextDocument textDocument)
+        private void TextDocumentActionOccurred(object sender, TextDocumentFileActionEventArgs e)
         {
-            document.DocumentRenamed += DocumentRenamed;
-            textDocument.TextBuffer.Properties.AddProperty(typeof(IDocument), document);
+            if (e.FileActionType != FileActionTypes.DocumentRenamed)
+                return;
 
-            _documents.Add(document.Path, document);
-        }
-
-        private void TextDocumentDisposed(object sender, TextDocumentEventArgs e)
-        {
-            if (_documents.TryGetValue(e.TextDocument.FilePath, out var document))
+            foreach (var path in _documents.Keys)
             {
-                document.DocumentRenamed -= DocumentRenamed;
-                document.Dispose();
-                _documents.Remove(e.TextDocument.FilePath);
-                DocumentDisposed?.Invoke(document);
+                var documentPair = _documents[path];
+                if (!documentPair.Key.Equals(sender))
+                    continue;
+
+                _documents[e.FilePath] = documentPair;
+                _documents.Remove(path);
+                break;
             }
         }
 
-        private void DocumentRenamed(IDocument document, string oldPath, string newPath)
+        private void OnDocumentDestroy(IDocument document)
         {
-            if (_documents.Remove(oldPath))
-                _documents.Add(newPath, document);
+            if (!_documents.TryGetValue(document.Path, out var documentPair))
+                return;
+
+            documentPair.Key.FileActionOccurred -= TextDocumentActionOccurred;
+            _documents.Remove(document.Path);
+            DocumentDisposed?.Invoke(document);
         }
 
-        private void OnChangeActivatedWindow(Window GotFocus, Window LostFocus)
+        private void OnChangeActivatedWindow(Window gotFocus, Window _)
```

Comment:
From the context of using, it seems to me that this should be called `OnWindowActivatedChanged`?